phoenix-core/src/main/java/org/apache/phoenix/jdbc/ConnectionInfo.java

```diff
@@ -0,0 +1,546 @@
+        return create(url, conf, props, info);
+    }
+
+    public static ConnectionInfo createNoLogin(String url, Configuration configuration,
+            ReadOnlyProps props, Properties info) throws SQLException {
+        return create(url, configuration, props, info, true);
+    }
+
+    public static ConnectionInfo create(String url, Configuration configuration,
+            ReadOnlyProps props, Properties info) throws SQLException {
+        return create(url, configuration, props, info, false);
+    }
+
+    public static ConnectionInfo create(String url, Configuration configuration,
+            ReadOnlyProps props, Properties info, boolean doNotLogin) throws SQLException {
+        // registry-independent URL preprocessing
+        url = url == null ? "" : url;
+        url = unescape(url);
+
+        // Assume missing prefix
+        if (url.isEmpty()) {
+            url = PhoenixRuntime.JDBC_PROTOCOL;
+        }
+        if (!url.startsWith(PhoenixRuntime.JDBC_PROTOCOL)) {
+            url = PhoenixRuntime.JDBC_PROTOCOL + PhoenixRuntime.JDBC_PROTOCOL_SEPARATOR + url;
+        }
+
+        if (configuration == null) {
+            configuration = HBaseFactoryProvider.getConfigurationFactory().getConfiguration();
+        }
+
+        Builder builder;
+
+        if (url.toLowerCase().startsWith(PhoenixRuntime.JDBC_PROTOCOL_ZK)) {
+            builder = new ZKConnectionInfo.Builder(url, configuration, props, info);
+        } else if (url.toLowerCase().startsWith(PhoenixRuntime.JDBC_PROTOCOL_MASTER)) {
+            builder = new MasterConnectionInfo.Builder(url, configuration, props, info);
+        } else if (url.toLowerCase().startsWith(PhoenixRuntime.JDBC_PROTOCOL_RPC)) {
+            builder = new RPCConnectionInfo.Builder(url, configuration, props, info);
+        } else if (url.toLowerCase().startsWith(PhoenixRuntime.JDBC_PROTOCOL)) {
+            // The generic protocol was specified. Try to Determine the protocol from the config
+            if (MasterConnectionInfo.isMaster(configuration)) {
+                builder = new MasterConnectionInfo.Builder(url, configuration, props, info);
+            } else if (RPCConnectionInfo.isRPC(configuration)) {
+                builder = new RPCConnectionInfo.Builder(url, configuration, props, info);
+            } else if (ZKConnectionInfo.isZK(configuration)) {
+                builder = new ZKConnectionInfo.Builder(url, configuration, props, info);
+            } else {
+                // No registry class set in config. Use version-dependent default
+                if (VersionInfo.getMajorVersion(VersionInfo.getVersion()) >= 3) {
+                    builder = new RPCConnectionInfo.Builder(url, configuration, props, info);
+                } else {
+                    builder = new ZKConnectionInfo.Builder(url, configuration, props, info);
+                }
+            }
+        } else {
+            throw getMalFormedUrlException(url);
+        }
```

Comment:
I wonder how best we can document this and what would be the right place

Reply:
I have added very long release note to the ticket as a stopgap, but we're gonna have to update the web page for sure.

Reply:
I'm gonna do that after the patch has landed.
Probably a dedicated page would be the best, then we could link to it from every place where the URLs are referred.
I would also like to emphasize in the docs the best solution is just to rely the defaults in hbase-site.xml